### DISCUSSION TITLE
Forcing positive solution on entire problem domain

### DISCUSSION THREAD

**Discussion by imilos** [lululxvi/deepxde#778] [General]
I have a 1D time-dependent problem and I'm trying to find a way to force the solution `n(x,t)` to be positive over the entire geomtime domain. I tried something like:

```
def bc_func1(inputs, outputs, X):
    return 1 - tf.sign(outputs)

def boundary_all(x, on_boundary):
    return True

ic2 = dde.icbc.OperatorBC(geomtime, bc_func1, boundary_all)

```
But it seems that it is the way. Which one is? Thanks in advance!

**Comment by imilos** [comment]
Since the condition `n(x,t)>0` applies to the whole geomtime domain, maybe it is better to specify it within the PDE() function somehow?

**Reply by forxltk** [reply]
Yes, or you can use a surrogate model by `apply_output_transform`.

**Comment by imilos** [comment]
Many thanks! I finally did it within `PDE()` function itself. However, it seems that my problem (Huxley muscle model) cannot be approached without weighted loss. In SCIANN library, I used to use something like `adaptive_weights={'method':'NTK', 'freq':50}`. Is there a similar feature in DeepXDE and a comprehensive example?

**Reply by forxltk** [reply]
See https://deepxde.readthedocs.io/en/latest/demos/pinn_forward/ode.2nd.html for `loss weights`. As I know, the adaptive weights algorithm has not been implemented in DeepXDE, you need to set the weights manually.

**Comment by athtareq** [comment]
Hello @imilos, can you please show the code that you used to force the solution to be positive ?   

A workaround I've tried is 

```
def output_transform(x, y):
  return tf.sqrt(y**2) 
net.apply_output_transform(output_transform)
```
and it works most of the time but the execution time is very large. I wonder how you did it ? Thanks

**Reply by imilos** [reply]
Hi @athtareq. Something like this worked for me:
```
def pde(x, n):
    dn_dt = dde.grad.jacobian(n, x, i=0, j=1)
    loss = dn_dt - (1.0-n) * f(x[:,0:1],a) + n*g(x[:,0:1])
    # Assure positive
    return loss + n*(1-tf.sign(n))

```

**Reply by athtareq** [reply]
@imilos that is a very clever approach, thank you !
